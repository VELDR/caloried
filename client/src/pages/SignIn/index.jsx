import PropTypes from 'prop-types';
import { useNavigate } from 'react-router-dom';
import { useState } from 'react';
import { useDispatch } from 'react-redux';
import loginIllustration from '@static/images/login-illustration.svg';
import adminIllustration from '@static/images/admin-illustration.svg';
import { FormattedMessage, injectIntl } from 'react-intl';

import FormInput from '@components/ui/FormInput';
import { AlternateEmail, Key } from '@mui/icons-material';
import { adminLogin, forgotPassword, login } from '@containers/Client/actions';

import classes from './style.module.scss';

const SignIn = ({ intl: { formatMessage } }) => {
  const dispatch = useDispatch();
  const navigate = useNavigate();
  const [email, setEmail] = useState('');
  const [password, setPassword] = useState('');
  const [isAdmin, setIsAdmin] = useState(false);

  const handleLogin = () => {
    const data = { email, password };
    if (isAdmin) {
      dispatch(
        adminLogin(data, () => {
          navigate('/admin');
        })
      );
    } else {
      dispatch(
        login(data, () => {
          navigate('/diary');
        })
      );
    }
  };

  const handleAdminChange = (e) => {
    setIsAdmin(e.target.checked);
  };

  const handleForgotPassword = () => {
    if (email) {
      dispatch(forgotPassword({ email }));
    }
  };

  return (
    <div className={classes.container} data-testid="sign-in">
      <div className={isAdmin ? `${classes.hero} ${classes.admin}` : `${classes.hero}`}>
        <div className={classes.hero__title}>
          {isAdmin ? formatMessage({ id: 'app_admin_sign_in_title' }) : formatMessage({ id: 'app_user_sign_in_title' })}
        </div>
        <div className={classes.hero__description}>
          {isAdmin
            ? formatMessage({ id: 'app_admin_sign_in_description' })
            : formatMessage({ id: 'app_user_sign_in_description' })}
        </div>
        <img src={isAdmin ? adminIllustration : loginIllustration} alt="Food Illustration" />
      </div>
      <div className={classes.signIn}>
        <div className={classes.signIn__title}>
          <FormattedMessage id="app_sign_in" />
        </div>
        <div className={classes.form}>
          <FormInput
            label={formatMessage({ id: 'app_email' })}
            icon={<AlternateEmail />}
            type="email"
            name="email"
            onChange={(e) => setEmail(e.target.value)}
          />
          <div>
            <FormInput
              label={formatMessage({ id: 'app_password' })}
              icon={<Key />}
              type="password"
              name="password"
              onChange={(e) => setPassword(e.target.value)}
            />
            {!isAdmin && (
              <div className={classes.form__forgot} onClick={handleForgotPassword}>
                <FormattedMessage id="app_forgot_password" />
              </div>
            )}
          </div>
          <div
            className={isAdmin ? `${classes.form__button} ${classes.admin}` : classes.form__button}
            onClick={handleLogin}
            data-testid="sign-in-button"
          >
            <FormattedMessage id="app_sign_in" />
          </div>
          {!isAdmin && (
            <div className={classes.form__signUp}>
              <FormattedMessage id="app_dont_have_account" />{' '}
              <a href="/sign-up">
                <FormattedMessage id="app_sign_up" />
              </a>
            </div>
          )}
          <div className={classes.option}>
            <div
              className={isAdmin ? classes.option__text : `${classes.option__selected} ${classes.option__text}`}
              data-testid="option-role-user"
            >
              User
            </div>

            <label htmlFor="switch" className={classes.option__toggle}>
              <input
                type="checkbox"
                id="switch"
                className={classes.option__checkbox}
                checked={isAdmin}
                onChange={handleAdminChange}
                data-testid="role-option-switch"
              />
              <span className={classes.option__slider} id="round" />
            </label>
            <div
              className={isAdmin ? `${classes.option__selected} ${classes.option__text}` : classes.option__text}
              data-testid="option-role-admin"
            >
              Admin
            </div>
          </div>
        </div>
      </div>
    </div>
  );
};

SignIn.propTypes = {
  intl: PropTypes.object,
};

export default injectIntl(SignIn);
