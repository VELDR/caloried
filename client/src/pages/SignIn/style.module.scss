@import '@styles/base';

.container {
  @include flex-center;

  // background-color: var(--color-bg-secondary);
  background-color: hsla(var(--hue), var(--saturation), var(--lightness), 0.65);
  border-radius: 2rem;
  border: 1px solid var(--color-text-secondary);
  box-sizing: border-box;
  @include box(20rem, 70vh);
  position: relative;
  text-align: center;
  @include apply-at-md-size {
    @include box(80vw, 70vh);
    overflow: hidden;
  }

  .hero {
    display: none;
    @include apply-at-md-size {
      @include flex-center-column;
      @include box(60%, 100%);
      background-color: var(--color-secondary);
      padding: 1rem;
      box-sizing: border-box;
    }

    &__title {
      font-size: $font-size-larger;
      font-weight: 700;
    }
    &__description {
      font-size: $font-size-small;
    }
    img {
      @include box(75%);
    }
  }
  .admin {
    background-color: var(--color-text-secondary);
  }

  .signIn {
    @include flex-center-column(1rem);
    @include apply-at-md-size {
      width: 40%;
    }

    &__title {
      font-size: $font-size-larger;
      font-weight: 700;
    }
    .form {
      @include flex-column(1rem);
      &__forgot {
        cursor: pointer;
        font-size: $font-size-smaller;
        text-align: right;
        &:hover {
          color: var(--color-text-secondary);
          text-decoration: underline;
        }
      }
      &__button {
        @include primary-button;
      }
      .admin {
        background-color: var(--color-text-secondary);
      }

      &__signUp {
        font-size: $font-size-smaller;
        a {
          color: var(--color-accent);
        }
      }
    }
  }
}

.option {
  @include flex-center;
  background-color: var(--color-bg-secondary);
  border-radius: 5px;
  gap: 2rem;
  padding: 0.25rem;
  color: var(--color-text-secondary);
  @include apply-at-lg-size {
    font-size: 1rem;
  }
  &__toggle {
    position: relative;
    display: inline-block;
    width: 35px;
    height: 20px;
  }
  &__checkbox {
    opacity: 0;
    @include box(0);

    &:checked + .option__slider::before {
      transform: translateX(14px);
    }
  }
  &__slider {
    position: absolute;
    cursor: pointer;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    background-color: var(--color-secondary);
    transition: 0.4s cubic-bezier(0, 1, 0.5, 1);
    border-radius: 34px;
    &::before {
      position: absolute;
      content: '';
      @include box(12px);
      left: 4px;
      bottom: 4px;
      background-color: white;
      transition: 0.4s cubic-bezier(0, 1, 0.5, 1);
      border-radius: 50%;
    }
  }
  &__text {
    font-size: $font-size-small;
  }
  &__selected {
    color: var(--color-text-primary);
    font-weight: 700;
  }
}
