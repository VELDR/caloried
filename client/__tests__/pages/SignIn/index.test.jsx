import SignIn from '@pages/SignIn';
import { fireEvent, render } from '@utils/testHelper';

const mockNavigate = jest.fn();

jest.mock('react-router-dom', () => ({
  useNavigate: () => mockNavigate,
}));

let wrapper;

const mockProps = {
  intl: {
    formatMessage: jest.fn(),
  },
};

beforeEach(() => {
  wrapper = render(<SignIn {...mockProps} />);
});

describe('SignIn Component', () => {
  it('should render the component correctly', () => {
    const { getByTestId } = wrapper;

    expect(getByTestId('sign-in')).toBeInTheDocument();
  });

  it('should update input values', () => {
    const { getByTestId } = wrapper;
    const emailInput = getByTestId('form-input-email');
    const passwordInput = getByTestId('form-input-password');

    expect(emailInput).toBeInTheDocument();
    expect(passwordInput).toBeInTheDocument();

    fireEvent.change(emailInput, { target: { value: 'test@example.com' } });
    fireEvent.change(passwordInput, { target: { value: 'password123' } });

    expect(emailInput.value).toBe('test@example.com');
    expect(passwordInput.value).toBe('password123');
  });

  it('should toggle between user and admin mode when the switch is clicked', () => {
    const { getByTestId } = wrapper;

    expect(getByTestId('option-role-user')).toHaveClass('option__selected');
    expect(getByTestId('option-role-admin')).not.toHaveClass('option__selected');

    fireEvent.click(getByTestId('role-option-switch'));

    expect(getByTestId('option-role-admin')).toHaveClass('option__selected');
    expect(getByTestId('option-role-user')).not.toHaveClass('option__selected');
  });

  it('should match with snapshot', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
