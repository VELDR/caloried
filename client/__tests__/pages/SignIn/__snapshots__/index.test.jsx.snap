// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SignIn Component should match with snapshot 1`] = `
{
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="container"
        data-testid="sign-in"
      >
        <div
          class="hero"
        >
          <div
            class="hero__title"
          >
            Kembali untuk Update Nutrisi Anda?
          </div>
          <div
            class="hero__description"
          >
            Masuk kembali untuk menjaga keseimbangan nutrisi Anda
          </div>
          <img
            alt="Food Illustration"
          />
        </div>
        <div
          class="signIn"
        >
          <div
            class="signIn__title"
          >
            Masuk
          </div>
          <div
            class="form"
          >
            <div
              class="form"
            >
              <div
                class="form__label"
              >
                Email
              </div>
              <div
                class="form__input"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="AlternateEmailIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10h5v-2h-5c-4.34 0-8-3.66-8-8s3.66-8 8-8 8 3.66 8 8v1.43c0 .79-.71 1.57-1.5 1.57s-1.5-.78-1.5-1.57V12c0-2.76-2.24-5-5-5s-5 2.24-5 5 2.24 5 5 5c1.38 0 2.64-.56 3.54-1.47.65.89 1.77 1.47 2.96 1.47 1.97 0 3.5-1.6 3.5-3.57V12c0-5.52-4.48-10-10-10zm0 13c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
                  />
                </svg>
                <input
                  data-testid="form-input-email"
                  name="email"
                  type="email"
                  value=""
                />
              </div>
            </div>
            <div>
              <div
                class="form"
              >
                <div
                  class="form__label"
                >
                  Kata Sandi
                </div>
                <div
                  class="form__input"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="KeyIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M21 10h-8.35C11.83 7.67 9.61 6 7 6c-3.31 0-6 2.69-6 6s2.69 6 6 6c2.61 0 4.83-1.67 5.65-4H13l2 2 2-2 2 2 4-4.04L21 10zM7 15c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z"
                    />
                  </svg>
                  <input
                    data-testid="form-input-password"
                    name="password"
                    type="password"
                    value=""
                  />
                  <div
                    class="eyeIcon"
                  >
                    <svg
                      aria-hidden="true"
                      class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                      data-testid="VisibilityIcon"
                      focusable="false"
                      viewBox="0 0 24 24"
                    >
                      <path
                        d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                      />
                    </svg>
                  </div>
                </div>
              </div>
              <div
                class="form__forgot"
              >
                Lupa Kata Sandi?
              </div>
            </div>
            <div
              class="form__button"
              data-testid="sign-in-button"
            >
              Masuk
            </div>
            <div
              class="form__signUp"
            >
              Tidak punya akun?
               
              <a
                href="/sign-up"
              >
                Daftar
              </a>
            </div>
            <div
              class="option"
            >
              <div
                class="option__selected option__text"
                data-testid="option-role-user"
              >
                User
              </div>
              <label
                class="option__toggle"
                for="switch"
              >
                <input
                  class="option__checkbox"
                  data-testid="role-option-switch"
                  id="switch"
                  type="checkbox"
                />
                <span
                  class="option__slider"
                  id="round"
                />
              </label>
              <div
                class="option__text"
                data-testid="option-role-admin"
              >
                Admin
              </div>
            </div>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="container"
      data-testid="sign-in"
    >
      <div
        class="hero"
      >
        <div
          class="hero__title"
        >
          Kembali untuk Update Nutrisi Anda?
        </div>
        <div
          class="hero__description"
        >
          Masuk kembali untuk menjaga keseimbangan nutrisi Anda
        </div>
        <img
          alt="Food Illustration"
        />
      </div>
      <div
        class="signIn"
      >
        <div
          class="signIn__title"
        >
          Masuk
        </div>
        <div
          class="form"
        >
          <div
            class="form"
          >
            <div
              class="form__label"
            >
              Email
            </div>
            <div
              class="form__input"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="AlternateEmailIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M12 2C6.48 2 2 6.48 2 12s4.48 10 10 10h5v-2h-5c-4.34 0-8-3.66-8-8s3.66-8 8-8 8 3.66 8 8v1.43c0 .79-.71 1.57-1.5 1.57s-1.5-.78-1.5-1.57V12c0-2.76-2.24-5-5-5s-5 2.24-5 5 2.24 5 5 5c1.38 0 2.64-.56 3.54-1.47.65.89 1.77 1.47 2.96 1.47 1.97 0 3.5-1.6 3.5-3.57V12c0-5.52-4.48-10-10-10zm0 13c-1.66 0-3-1.34-3-3s1.34-3 3-3 3 1.34 3 3-1.34 3-3 3z"
                />
              </svg>
              <input
                data-testid="form-input-email"
                name="email"
                type="email"
                value=""
              />
            </div>
          </div>
          <div>
            <div
              class="form"
            >
              <div
                class="form__label"
              >
                Kata Sandi
              </div>
              <div
                class="form__input"
              >
                <svg
                  aria-hidden="true"
                  class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                  data-testid="KeyIcon"
                  focusable="false"
                  viewBox="0 0 24 24"
                >
                  <path
                    d="M21 10h-8.35C11.83 7.67 9.61 6 7 6c-3.31 0-6 2.69-6 6s2.69 6 6 6c2.61 0 4.83-1.67 5.65-4H13l2 2 2-2 2 2 4-4.04L21 10zM7 15c-1.65 0-3-1.35-3-3s1.35-3 3-3 3 1.35 3 3-1.35 3-3 3z"
                  />
                </svg>
                <input
                  data-testid="form-input-password"
                  name="password"
                  type="password"
                  value=""
                />
                <div
                  class="eyeIcon"
                >
                  <svg
                    aria-hidden="true"
                    class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                    data-testid="VisibilityIcon"
                    focusable="false"
                    viewBox="0 0 24 24"
                  >
                    <path
                      d="M12 4.5C7 4.5 2.73 7.61 1 12c1.73 4.39 6 7.5 11 7.5s9.27-3.11 11-7.5c-1.73-4.39-6-7.5-11-7.5zM12 17c-2.76 0-5-2.24-5-5s2.24-5 5-5 5 2.24 5 5-2.24 5-5 5zm0-8c-1.66 0-3 1.34-3 3s1.34 3 3 3 3-1.34 3-3-1.34-3-3-3z"
                    />
                  </svg>
                </div>
              </div>
            </div>
            <div
              class="form__forgot"
            >
              Lupa Kata Sandi?
            </div>
          </div>
          <div
            class="form__button"
            data-testid="sign-in-button"
          >
            Masuk
          </div>
          <div
            class="form__signUp"
          >
            Tidak punya akun?
             
            <a
              href="/sign-up"
            >
              Daftar
            </a>
          </div>
          <div
            class="option"
          >
            <div
              class="option__selected option__text"
              data-testid="option-role-user"
            >
              User
            </div>
            <label
              class="option__toggle"
              for="switch"
            >
              <input
                class="option__checkbox"
                data-testid="role-option-switch"
                id="switch"
                type="checkbox"
              />
              <span
                class="option__slider"
                id="round"
              />
            </label>
            <div
              class="option__text"
              data-testid="option-role-admin"
            >
              Admin
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
